File: src/scraper.rs
use scraper::{Html, Selector};
use serde_json::Value;

use crate::http_client::CurlClient;
use crate::models::MediaItem;

pub struct Scraper {
    pub base_url: String,
    pub client: CurlClient,
}

impl Scraper {
    pub fn new(base_url: Option<String>) -> Self {
        let base = base_url.unwrap_or_else(|| "https://z2.idlixku.com".to_string());
        Self {
            base_url: base,
            client: CurlClient::new(),
        }
    }

    /// Search content via IDLIX JSON API endpoint
    pub async fn search_content(&self, query: &str) -> Result<Vec<MediaItem>, String> {
        let q_clean = query.trim();
        if q_clean.is_empty() {
            return Ok(Vec::new());
        }

        let api_url = format!("{}/api/search?q={}", self.base_url, urlencoding::encode(q_clean));
        let referer = format!("{}/", self.base_url);

        let body = self
            .client
            .get(&api_url, Some(&referer), true)
            .await
            .map_err(|e| format!("Search request failed: {e}"))?;

        let data: Value = serde_json::from_str(&body)
            .map_err(|e| format!("Failed to parse search JSON: {e} (body snippet: {})", &body.chars().take(200).collect::<String>()))?;

        let results = data["results"]
            .as_array()
            .ok_or_else(|| "Missing results array in response".to_string())?;

        let mut items = Vec::new();
        for item in results {
            let title = item["title"]
                .as_str()
                .or_else(|| item["name"].as_str())
                .unwrap_or("Unknown")
                .to_string();

            let slug = item["slug"].as_str().unwrap_or_default().to_string();
            let content_type = item["contentType"].as_str().unwrap_or("");

            let is_tv = content_type == "tv_series"
                || content_type == "series"
                || content_type == "tvshows"
                || slug.contains("series/")
                || slug.contains("tvshows/");

            let media_type = if is_tv { "TV Series".to_string() } else { "Movie".to_string() };
            let url = if is_tv {
                format!("{}/series/{}", self.base_url, slug)
            } else {
                format!("{}/movie/{}", self.base_url, slug)
            };

            let rel_date = item["releaseDate"]
                .as_str()
                .or_else(|| item["firstAirDate"].as_str())
                .unwrap_or("");
            let year = if rel_date.len() >= 4 {
                Some(rel_date[0..4].to_string())
            } else {
                None
            };

            let rating = if let Some(vote) = item["voteAverage"].as_f64() {
                format!("{:.1}", vote)
            } else {
                "N/A".to_string()
            };

            let raw_poster = item["posterPath"].as_str().or_else(|| item["poster"].as_str()).unwrap_or("");
            let poster = if raw_poster.is_empty() {
                String::new()
            } else if raw_poster.starts_with("http") {
                raw_poster.to_string()
            } else {
                format!("https://image.tmdb.org/t/p/w500{}", raw_poster)
            };

            items.push(MediaItem {
                title,
                url,
                slug,
                rating,
                media_type,
                poster,
                year,
            });
        }

        Ok(items)
    }

    /// Fetches Featured / Popular items from home page JSON-LD schema & resolves metadata
    pub async fn fetch_featured(&self) -> Result<Vec<MediaItem>, String> {
        let html = match self.client.get(&self.base_url, None, false).await {
            Ok(body) => body,
            Err(_) => return self.search_content("2024").await,
        };

        let mut featured_targets = Vec::new();

        // 1. Parse JSON-LD structured data from IDLIX Next.js Homepage
        {
            let document = Html::parse_document(&html);
            if let Ok(script_selector) = Selector::parse("script[type='application/ld+json']") {
                for script_el in document.select(&script_selector) {
                    let text = script_el.text().collect::<Vec<_>>().join("");
                    if let Ok(json_val) = serde_json::from_str::<Value>(&text) {
                        if let Some(arr) = json_val.as_array() {
                            for item in arr {
                                if item["@type"] == "ItemList" {
                                    if let Some(elements) = item["itemListElement"].as_array() {
                                        for el in elements {
                                            let name = el["name"].as_str().unwrap_or_default().to_string();
                                            let url = el["url"].as_str().unwrap_or_default().to_string();
                                            if !name.is_empty() && !url.is_empty() {
                                                let slug = url.trim_matches('/').split('/').last().unwrap_or("").to_string();
                                                let is_tv = url.contains("/series/");
                                                featured_targets.push((name, url, slug, is_tv));
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        // 2. Resolve metadata (Posters, ratings, release date) for all featured titles concurrently
        if !featured_targets.is_empty() {
            let mut join_handles = Vec::new();
            for (name, url, slug, is_tv) in featured_targets {
                let client = self.client.clone();
                let base_url = self.base_url.clone();
                join_handles.push(tokio::spawn(async move {
                    let scraper = Scraper { base_url, client };
                    if let Ok(search_res) = scraper.search_content(&name).await {
                        if let Some(found) = search_res.into_iter().find(|r| r.slug == slug || r.title.eq_ignore_ascii_case(&name)) {
                            return found;
                        }
                    }
                    MediaItem {
                        title: name,
                        url,
                        slug,
                        rating: "Featured".to_string(),
                        media_type: if is_tv { "TV Series".to_string() } else { "Movie".to_string() },
                        poster: String::new(),
                        year: None,
                    }
                }));
            }

            let mut resolved_items = Vec::new();
            for handle in join_handles {
                if let Ok(item) = handle.await {
                    resolved_items.push(item);
                }
            }

            if !resolved_items.is_empty() {
                return Ok(resolved_items);
            }
        }

        self.search_content("2024").await
    }
}

File: internal/scraper/scraper.go
package scraper

import (
	"encoding/json"
	"fmt"
	"net/url"
	"strings"
	"sync"

	"github.com/PuerkitoBio/goquery"
	"github.com/iamnotpirates/idlixdownloader/internal/httpclient"
	"github.com/iamnotpirates/idlixdownloader/internal/models"
)

type Scraper struct {
	BaseURL string
	client  *httpclient.CurlClient
}

func New(baseURL string, client *httpclient.CurlClient) *Scraper {
	if baseURL == "" {
		baseURL = "https://z2.idlixku.com"
	}
	if client == nil {
		client = httpclient.New()
	}
	return &Scraper{
		BaseURL: strings.TrimRight(baseURL, "/"),
		client:  client,
	}
}

func (s *Scraper) SearchContent(query string) ([]models.MediaItem, error) {
	return s.SearchContentWithType(query, "")
}

func (s *Scraper) SearchContentWithType(query, typeFilter string) ([]models.MediaItem, error) {
	qClean := strings.TrimSpace(query)
	if qClean == "" {
		return nil, nil
	}

	apiURL := fmt.Sprintf("%s/api/search?q=%s", s.BaseURL, url.QueryEscape(qClean))
	if typeFilter != "" {
		validType := typeFilter
		if typeFilter == "series" || typeFilter == "tv" {
			validType = "tv_series"
		}
		apiURL = fmt.Sprintf("%s/api/search?q=%s&type=%s", s.BaseURL, url.QueryEscape(qClean), url.QueryEscape(validType))
	}
	referer := fmt.Sprintf("%s/", s.BaseURL)

	body, err := s.client.Get(apiURL, referer, true)
	if err != nil {
		return nil, fmt.Errorf("search request failed: %w", err)
	}

	var data struct {
		Results []struct {
			Title        string   `json:"title"`
			Name         string   `json:"name"`
			Slug         string   `json:"slug"`
			ContentType  string   `json:"contentType"`
			ReleaseDate  string   `json:"releaseDate"`
			FirstAirDate string   `json:"firstAirDate"`
			VoteAverage  *float64 `json:"voteAverage"`
			PosterPath   string   `json:"posterPath"`
			Poster       string   `json:"poster"`
		} `json:"results"`
	}

	if err := json.Unmarshal([]byte(body), &data); err != nil {
		return nil, fmt.Errorf("failed to parse search JSON: %w", err)
	}

	var items []models.MediaItem
	for _, item := range data.Results {
		title := item.Title
		if title == "" {
			title = item.Name
		}
		if title == "" {
			title = "Unknown"
		}

		isTV := item.ContentType == "tv_series" ||
			item.ContentType == "series" ||
			item.ContentType == "tvshows" ||
			strings.Contains(item.Slug, "series/") ||
			strings.Contains(item.Slug, "tvshows/")

		if typeFilter == "movie" && isTV {
			continue
		}
		if (typeFilter == "tv_series" || typeFilter == "series" || typeFilter == "tv") && !isTV {
			continue
		}

		mediaType := "Movie"
		itemURL := fmt.Sprintf("%s/movie/%s", s.BaseURL, item.Slug)
		if isTV {
			mediaType = "TV Series"
			itemURL = fmt.Sprintf("%s/series/%s", s.BaseURL, item.Slug)
		}

		relDate := item.ReleaseDate
		if relDate == "" {
			relDate = item.FirstAirDate
		}
		var year *string
		if len(relDate) >= 4 {
			y := relDate[:4]
			year = &y
		}

		rating := "N/A"
		if item.VoteAverage != nil {
			rating = fmt.Sprintf("%.1f", *item.VoteAverage)
		}

		rawPoster := item.PosterPath
		if rawPoster == "" {
			rawPoster = item.Poster
		}
		poster := ""
		if rawPoster != "" {
			if strings.HasPrefix(rawPoster, "http") {
				poster = rawPoster
			} else {
				poster = fmt.Sprintf("https://image.tmdb.org/t/p/w500%s", rawPoster)
			}
		}

		items = append(items, models.MediaItem{
			Title:     title,
			URL:       itemURL,
			Slug:      item.Slug,
			Rating:    rating,
			MediaType: mediaType,
			Poster:    poster,
			Year:      year,
		})
	}

	return items, nil
}

func (s *Scraper) FetchFeatured() ([]models.MediaItem, error) {
	html, err := s.client.Get(s.BaseURL, "", false)
	if err != nil {
		// Fallback to default search
		return s.SearchContent("2024")
	}

	type target struct {
		name string
		url  string
		slug string
		isTV bool
	}

	var targets []target

	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err == nil {
		doc.Find("script[type='application/ld+json']").Each(func(i int, sel *goquery.Selection) {
			text := sel.Text()
			var jsonVals []map[string]any
			if err := json.Unmarshal([]byte(text), &jsonVals); err == nil {
				for _, obj := range jsonVals {
					if obj["@type"] == "ItemList" {
						if elems, ok := obj["itemListElement"].([]any); ok {
							for _, el := range elems {
								if m, ok := el.(map[string]any); ok {
									name, _ := m["name"].(string)
									u, _ := m["url"].(string)
									if name != "" && u != "" {
										parts := strings.Split(strings.Trim(u, "/"), "/")
										slug := parts[len(parts)-1]
										isTV := strings.Contains(u, "/series/")
										targets = append(targets, target{name: name, url: u, slug: slug, isTV: isTV})
									}
								}
							}
						}
					}
				}
			}
		})
	}

	if len(targets) == 0 {
		return s.SearchContent("2024")
	}

	var wg sync.WaitGroup
	items := make([]models.MediaItem, len(targets))

	for i, t := range targets {
		wg.Add(1)
		go func(idx int, tgt target) {
			defer wg.Done()
			res, err := s.SearchContent(tgt.name)
			if err == nil {
				for _, r := range res {
					if r.Slug == tgt.slug || strings.EqualFold(r.Title, tgt.name) {
						items[idx] = r
						return
					}
				}
			}
			mType := "Movie"
			if tgt.isTV {
				mType = "TV Series"
			}
			items[idx] = models.MediaItem{
				Title:     tgt.name,
				URL:       tgt.url,
				Slug:      tgt.slug,
				Rating:    "Featured",
				MediaType: mType,
			}
		}(i, t)
	}
	wg.Wait()

	var result []models.MediaItem
	for _, it := range items {
		if it.Title != "" {
			result = append(result, it)
		}
	}

	return result, nil
}
